File: easy/2351. First Letter to Appear Twice.py
# 2351. First Letter to Appear Twice
# https://leetcode.com/contest/weekly-contest-303/problems/first-letter-to-appear-twice/

# 1
class Solution:
  def repeatedCharacter(self, s: str) -> str:
    letters = [0] * 26

    for i in s:
      idx = ord(i) - ord('a')
      if letters[idx] == 1:
        return i
      letters[idx] += 1

# 2
class Solution:
  def repeatedCharacter(self, s: str) -> str:
    hashSet = set()

    for ch in s:
      if ch in hashSet:
        return ch
      hashSet.add(ch)

File: easy/2351. First Letter to Appear Twice.js
// 2351. First Letter to Appear Twice

// https://leetcode.com/contest/weekly-contest-303/problems/first-letter-to-appear-twice/

// #1
const repeatedCharacter = function(s) {
  const set = new Set()

  for (let ch of s) {
    if (set.has(ch)) {
      return ch
    }
    set.add(ch)
  }
};

// #2
const repeatedCharacter2 = function(s) {
  const hash = Array(26).fill(0)

  for (let ch of s) {
    const code = ch.charCodeAt()
    if (hash[code - 'a'.charCodeAt()] > 0) {
      return ch
    }
    hash[code - 'a'.charCodeAt()]++
  }
};

// #3
const repeatedCharacter3 = function(s) {
  const hash = Array(26).fill(0)

  for (let ch of s) {
    const idx = ch.charCodeAt(0) - 'a'.charCodeAt(0)
    if (hash[idx]) {
      return ch
    }
    hash[idx]++
  }
};